### DISCUSSION TITLE
tquic 如何使用hook进行明文传输

### DISCUSSION THREAD

**Discussion by lion117** [Tencent/tquic#323] [Q&A (问题和回答)]
# 背景
1. ubuntux64
2. libtquic.c 
3. 内网WIFI环境, 有一定概率丢包

# 需求描述
![image](https://github.com/Tencent/tquic/assets/3996029/d863d2a8-9a9f-4478-9c88-a48124dfff29)
阅读下列这篇文章
https://cloud.tencent.com/developer/article/1908416, 其中提到一种方法使用hook的方式进行伪加解密.  
在调试过程中用wireshark实时分析明文quic 能够很好的提高效率.  到生产环境再切换到密文

# 问题
在C++环境下如何使用文章中提到的hook机制进行明文传输呢?

**Comment by glime-ai** [comment]
在C++环境下使用文章中提到的hook机制进行明文传输，可以参考以下步骤：

1. **设置Keylog回调**：在客户端和服务器端设置TLS密钥日志回调，以便在Wireshark中解密QUIC流量。

2. **实现Keylog回调函数**：编写一个回调函数，将密钥日志写入文件。

3. **配置SSL_CTX**：在创建SSL_CTX时，设置密钥日志回调。

以下是一个示例代码，展示了如何在C++环境下实现上述步骤：

```cpp
#include <openssl/ssl.h>
#include <openssl/err.h>
#include <cstdio>

// 回调函数，用于将密钥日志写入文件
void keylog_callback(const SSL *ssl, const char *line) {
    static FILE *keylog_file = nullptr;
    if (!keylog_file) {
        keylog_file = fopen("quic_keylog.log", "w");
    }
    if (keylog_file) {
        fprintf(keylog_file, "%s\n", line);
        fflush(keylog_file);
    }
}

int main() {
    // 初始化OpenSSL
    SSL_library_init();
    SSL_load_error_strings();
    OpenSSL_add_ssl_algorithms();

    // 创建SSL_CTX
    SSL_CTX *ctx = SSL_CTX_new(TLS_method());
    if (!ctx) {
        fprintf(stderr, "Failed to create SSL_CTX\n");
        return 1;
    }

    // 设置密钥日志回调
    SSL_CTX_set_keylog_callback(ctx, keylog_callback);

    // 其他QUIC/TLS配置...

    // 清理
    SSL_CTX_free(ctx);
    EVP_cleanup();
    return 0;
}
```

在上述代码中，`keylog_callback`函数会将TLS密钥日志写入`quic_keylog.log`文件。然后，在Wireshark中，可以通过以下步骤加载该密钥日志文件以解密QUIC流量：

1. 打开Wireshark。
2. 依次点击`Edit` -> `Preferences` -> `Protocols` -> `TLS`。
3. 在`(Pre)-Master-Secret log filename`字段中，选择生成的`quic_keylog.log`文件。

这样，Wireshark就可以实时解密并分析QUIC流量了。

This comment was generated by [Glime](https://glimelab.ai/?utm_source=44520fce49ba43798485a628d6928e22&utm_medium=github).

**Reply by lion117** [reply]
通过sskey的解密是一种方法, 但没有明文来的快捷方便.

**Comment by iyangsj** [comment]
明文模式没有正式支持的计划，有需求的用户可以按需定制
（可参考 https://datatracker.ietf.org/doc/draft-banks-quic-disable-encryption/ ）

**Reply by iyangsj** [reply]
[TQUIC 1.3.0版本](https://github.com/Tencent/tquic/releases/tag/v1.3.0)已支持 https://datatracker.ietf.org/doc/draft-banks-quic-disable-encryption/
